File: Go/leetCode/maxSubArray.go
package leetcode

func MaxSubArray(nums []int) int {
	max := nums[0]
	curr := nums[0]

	for i := 1; i < len(nums); i++ {
		curr = Max(nums[i], curr+nums[i])
		max = Max(max, curr)

	}
	return max
}

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Python/algoritms/leetCode/maxSubArray.py
from typing import List


def maxSubArray(nums) -> int:

    max_sum = float('-inf')
    current_sum = 0  # Initialize current_sum as 0

    for num in nums:
        current_sum += num
        max_sum = max(max_sum, current_sum)

        # If current_sum becomes negative, reset it to 0
        if current_sum < 0:
            current_sum = 0

    return max_sum


nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print(maxSubArray(nums))
